### ISSUE TITLE
Open Tailwind code in Tailwind Playground

### ISSUE THREAD

**Issue by davestewart** [bernaferrari/FigmaToCode#114]
Whilst not possible to do directly, a simple Chrome extension could act as a bridge.

The steps would be as follows:

- Figma plugin opens new window, either the Tailwind Playground URL directly (or maybe a chrome extension page)
- the Tailwind HTML encoded in the URL
- the extension could (I think?) run javascript in the main page using the [`MAIN`](https://developer.chrome.com/docs/extensions/reference/api/scripting#type-ExecutionWorld) execution context:

```js
const url = new URL(location.href)
const html = url.searchParams.get('html')
if (html) {
  MonacoEditor.getModel().setValue(html)
}
```

Have done a quick experiment, and the code does set the Tailwind Playground code, would just have to confirm the extension could access the `MonacoEditor` global.

**Comment by bernaferrari** [comment]
Maybe tailwind playground is not doable, but codepen/codesandbox/replit/others would be doable?

**Comment by davestewart** [comment]
I've been thinking through this a bit more, and actually it might just be better to render Tailwind directly in the responsive preview. 

See also:

- #116

**Comment by bernaferrari** [comment]
If you know how, I would accept.. But I don't know how.

**Comment by davestewart** [comment]
I think it might be possible using the Play CDN:

- https://tailwindcss.com/docs/installation/play-cdn

For arbitrary rules, a custom config can be created injected (see further down page). 

I don't know if it's possible to run inside the Figma environment, but might be an option.

Have also seen some solutions where classes were generated, but it adds about 1.2MB to the size of the CSS. For arbitrary classes (size, color) could probably parse the markup and add styles dynamically.

Happy to take a look.

**Comment by bernaferrari** [comment]
Play cdn won't work because plugin has no internet connection. Could flex
that a bit,allow and see if it works well, but I don't know.. What do you
think? 1mb is probably too much for Figma. I don't know how to fix this in
a good way.

**Comment by davestewart** [comment]
I've had a few attempts to make this work and coming up short.

### Attempt 1: Grab the Tailwind 2 compiled source

This mainly works, but I know there's a few changes from v2 to v3, esp with colors.

![CleanShot 2024-07-15 at 14 07 30](https://github.com/user-attachments/assets/41399694-e693-46e7-9e43-61774b595201)

### Attempt 2: Compile Tailwind using Post CSS

I got Tailwind compiling with JIT via Post CSS in Node, but I can't get it to run in the Figma build env.

![CleanShot 2024-07-15 at 13 51 53](https://github.com/user-attachments/assets/5097c3e1-5523-4959-b3d3-b6f81cac7f15)

Seems to be problems with built-ins like FS which I can't seem to get around, even though I've tried multiple polyfills etc.

Going to pause effort on this for now.

### Attempt 3:

Not tried this yet, but think could be an option to:

- identify the finite classes needed, such as layout, default colours, etc
- leverage Tailwind's default config (see `packages/backend/node_modules/tailwindcss/stubs/config.full.js`)
- build this set of classes using some of the techniques above
- create additional arbitrary classes (size, colors) on the fly

I think for now – as I need to get back to work – I'm going to pause this, but think it's worth exploring.

**Comment by bernaferrari** [comment]
Seems the only way is via the web. Thanks for looking trough. Another
possibility would somehow be mapping Tailwind colors and attributes to html
in a way. But don't worry.

**Comment by bernaferrari** [comment]
I liked approach 3 because if we make html output Tailwind colors somehow, then it is very easy. Only code would be potentially ugly.

**Comment by davestewart** [comment]
OK. I propose the PR would look something like this:

Build time:

- I push a script with the PostCSS generation code
- you work on the permutations for possible classes
- final script outputs all these classes to a static file which is loaded in Plugin UI

Run time

- design regexp to grab arbitrary class names `(text|bg)-\[(.+?)\]`
- add functions to generate arbitrary classes
- process HTML code to grab tokens
- pass Tailwind code and additional Tailwind classes to Plugin UI from backend

I'm happy to take the Post CSS stuff if you want to take more of the client code?

**Comment by davestewart** [comment]
It looks like additional processing will also be needed for transparent colors, as not only can `/25` be passed as a token to the compiler, but passing all those transparency variants would balloon the file size.

I think we can get somewhere which will work, but indeed, Tailwind is tricky!
lf.

**Comment by davestewart** [comment]
Some additional tweaks needed, but do seem to be getting somewhere:

![CleanShot 2024-07-15 at 18 01 42](https://github.com/user-attachments/assets/a5e18183-bd90-44f4-a83d-6b11496c7da5)

**Comment by bernaferrari** [comment]
Yeah, I can help. Whatever you do, make it easy to update, because Tailwind 4 will come one day and it is going to be big.

So, are we kind of reverse-engineering tailwind to output html? Does it add 1mb or not?

**Comment by davestewart** [comment]
> So, are we kind of reverse-engineering tailwind to output html

Kind of:

- pre-compiling 90% of the base classes
- generating the remaining 10%

Suspect will need to generate from markup:

- arbitrary colors
- transparent colors

Have been tweaking classes in the devtools and it feels way better to be able to toggle TW classes than guess what they might be once they are copied and pasted elsewhere. Not perfect compared to an actual JIT compiler, but feels like it could certainly be better than the HTML stand-in.

> Does it add 1mb or not?

117K at the mo, which is manageable.

**Comment by davestewart** [comment]
Morning update...

Tailwind classes:

- have added all supported TW classes; TW size is now 178K
- could _potentially_ output as JSON, and just add classes as needed (vs adding the whole stylesheet)

Preview generation:

- am successfully sending TW to preview from backend
- have arbitrary colors, opacities and sizes (`w h p m px py mx my gap`) working

Here's how it's looking right now (outline just for debugging purposes):

![CleanShot 2024-07-16 at 13 13 04](https://github.com/user-attachments/assets/1515e636-483a-46e4-8d17-4319752667cb)

**Comment by bernaferrari** [comment]
Looks fantastic. Do you need any help?

**Comment by davestewart** [comment]
Probably for QA!

And, maybe to save the state of the preview frame. I'm a Vue not React dev.

**Comment by davestewart** [comment]
Question regarding left and top values.

In the tailwind builder, size values pass through the conversion tables / rounding function, but position values do not.

Just checking that this is intentional?

(I wonder if at some point in the future there might be more granular control over which properties get rounded).

**Comment by davestewart** [comment]
OK, so FINALLY got all the flexbox stuff working:

![CleanShot 2024-07-19 at 19 37 55](https://github.com/user-attachments/assets/e31ec768-1b1f-42f7-bf90-19b50e0e4023)

It's pretty cool; the new renderer now shows real Tailwind, with the rounding of sizes, colors, as well as user colors (variables).

The reason why user colors are showing as stripes above, is because I'm working to add Figma's `codeSyntax` property, which will mean we won't have to fall back to deriving TW names names from variable names, the user will be able to specify them in the UI:

![image](https://github.com/user-attachments/assets/62e44d35-b1cd-444c-a234-7b82c936eff1)

Overall... very nearly there!

**Comment by bernaferrari** [comment]
Amazing!

**Comment by bernaferrari** [comment]
> In the tailwind builder, size values pass through the conversion tables / rounding function, but position values do not.

I don't know lol I think it wasn't intentional, but maybe I found it easier that way, because you have `top-4 left-[17px]`, quickly becomes weird, could be that.. I don't remember. Feel free to fix it!

**Comment by davestewart** [comment]
With the initial work on variables done, I'll look to implement them for numbers soon, then the choice to use exact values or rounded values can be chosen per-property.
